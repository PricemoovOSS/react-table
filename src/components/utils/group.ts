import { clone, cloneDeep } from "lodash";
import { IColumns } from "../table";

export interface Group {
  id: string;
  label?: string;
  subGroups?: Group[];
}

export interface GroupBranche extends Group {
  size: number;
  subGroups?: GroupBranche[];
}

const DEFAULT_GROUP_ID = "default-group-id";

/**
 *
 * @param groups the list of raw groups
 * @returns the max depth of the groups, and all the branches of the groups
 */
export function getGroupBranches(groups: Group[] = []): {
  depth: number;
  groupBranches: Record<string, GroupBranche> | undefined;
} {
  if (groups.length === 0) {
    return { depth: 0, groupBranches: undefined };
  }

  const clonedGroups = cloneDeep(groups);
  // add the default group for columns without group
  clonedGroups.push({ id: DEFAULT_GROUP_ID, label: "" });
  const depth = equalizeGroupDepths(clonedGroups);

  const branches = groupsToBranches(clonedGroups);
  const groupBranches = branches.reduce((acc, branche) => {
    const leaves = getLeaves(branche);
    acc[leaves[0].id] = branche;
    return acc;
  }, {});
  return { depth, groupBranches };
}

/**
 * Each column is associated with a given branch.
 * Branches must be merged to reconstitute groups consistent with the visible columns.
 * @param branches the list of alll branches
 * @returns Groups with their sizes
 */
export function getMergedBranches(branches: GroupBranche[]): GroupBranche[] {
  let currentBranche: GroupBranche;
  const result: GroupBranche[] = [];
  branches.forEach((branche, index) => {
    if (!currentBranche || currentBranche.id !== branche.id) {
      if (currentBranche) {
        result.push(currentBranche);
      }
      currentBranche = branche;
    } else if (currentBranche) {
      currentBranche = mergeBranches(branche, currentBranche);
      currentBranche.size += branche.size;
    }
    if (index === branches.length - 1) {
      result.push(currentBranche);
    }
  });

  return result;
}

/**
 *
 * @param groupBranches raw branches
 * @param columns columns options
 * @param visibleColumnIndexes the current visible cell indexes
 * @returns the list of branches associated with each specified column
 */
export function getColumnBranches(
  groupBranches: Record<string, GroupBranche>,
  columns: IColumns = {},
  visibleColumnIndexes: number[] = []
): GroupBranche[] {
  return visibleColumnIndexes.reduce<GroupBranche[]>((acc, cellIndex) => {
    const column = columns[cellIndex];
    const groupId = column?.groupId || DEFAULT_GROUP_ID;
    const branche = groupBranches[groupId];
    if (branche) {
      acc.push(cloneDeep(branche));
    }
    return acc;
  }, []);
}

function equalizeGroupDepths(groups: Group[], depth = 0): number {
  const hasSubGroup = groups.some((group) => group.subGroups && group.subGroups.length > 0);
  if (hasSubGroup) {
    const subGroups: Group[] = [];
    groups.forEach((group) => {
      if (!group.subGroups || group.subGroups.length === 0) {
        group.subGroups = [{ id: group.id, label: "" }];
        group.id = `${group.id}-root-${depth}`;
      }
      subGroups.push(...(group.subGroups as Group[]));
    });
    return equalizeGroupDepths(subGroups, depth + 1);
  }
  return depth + 1;
}

function groupsToBranches(groups: Group[]): GroupBranche[] {
  return groups.reduce<GroupBranche[]>((acc, group) => {
    const subBranches = groupsToBranches(group.subGroups || []);
    if (subBranches.length > 0) {
      subBranches.forEach((subBranche) => {
        const newGroup = clone(group) as GroupBranche;
        newGroup.size = 1;
        newGroup.subGroups = [subBranche];
        acc.push(newGroup);
      });
    } else {
      const newGroup = clone(group) as GroupBranche;
      newGroup.size = 1;
      acc.push(newGroup);
    }

    return acc;
  }, []);
}

function getLeaves(group: GroupBranche): GroupBranche[] {
  if (!group.subGroups || group.subGroups.length === 0) {
    return [group];
  }
  return group.subGroups.reduce<GroupBranche[]>((acc, subGroup) => {
    acc.push(...getLeaves(subGroup));
    return acc;
  }, []);
}

function mergeBranches(source: GroupBranche, target: GroupBranche): GroupBranche {
  if (source.subGroups && source.subGroups.length > 0) {
    const sourceSubBranche = source.subGroups.shift() as GroupBranche;
    const targetSubBranche = target.subGroups?.pop() as GroupBranche;
    if (sourceSubBranche.id === targetSubBranche.id) {
      const mergedBranche = mergeBranches(sourceSubBranche, targetSubBranche);
      mergedBranche.size = sourceSubBranche.size + targetSubBranche.size;
      target.subGroups?.push(mergedBranche, ...source.subGroups);
    } else {
      target.subGroups?.push(targetSubBranche, sourceSubBranche, ...source.subGroups);
    }
  }
  return target;
}
