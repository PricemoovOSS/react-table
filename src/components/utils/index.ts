export * from "./table";
export * from "./table-selection";
export * from "./common";
export * from "./group";

export { default as shallowEqual } from "./shallowEqual";
