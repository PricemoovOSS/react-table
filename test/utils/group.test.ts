/// <reference path="../typings/tests-entry.d.ts" />

import { IColumns } from "../../src/components";
import { GroupBranch, getColumnBranches, getGroupBranches, getMergedBranches } from "../../src/components/utils";

const groups = [
  { id: "group1", label: "Group 1", subGroups: [{ id: "subgroup-1", label: "Sub-Group 1.1" }] },
  {
    id: "group2",
    label: "Group 2",
    subGroups: [
      { id: "subgroup-2", label: "Sub-Group 2.1" },
      {
        id: "subgroup-3",
        label: "Sub-Group 2.2",
        subGroups: [
          { id: "subgroup-4", label: "Sub-Group 2.2.1" },
          { id: "subgroup-5", label: "Sub-Group 2.2.2" },
        ],
      },
    ],
  },
];

const columns: IColumns = {
  0: { groupId: "subgroup-1" },
  1: { groupId: "subgroup-1" },
  3: { groupId: "subgroup-2" },
  4: { groupId: "subgroup-5" },
  5: { groupId: "subgroup-5" },
};

describe("Grou utils", () => {
  test("should return not empty branches list", () => {
    const { depth, groupBranches } = getGroupBranches(groups);
    expect(depth).toEqual(3);
    expect(groupBranches).toMatchSnapshot();
  });

  test("should return empty branches", () => {
    const { depth, groupBranches } = getGroupBranches();
    expect(depth).toEqual(0);
    expect(groupBranches).toBeUndefined();

    const { depth: emptyDepth, groupBranches: emptyBranches } = getGroupBranches([]);
    expect(emptyDepth).toEqual(0);
    expect(emptyBranches).toBeUndefined();
  });

  test("should return visible column branches", () => {
    const { groupBranches } = getGroupBranches(groups);
    const result = getColumnBranches(groupBranches as Record<string, GroupBranch>, columns, [0, 1, 2, 3, 4, 5, 6]);
    expect(result).toMatchSnapshot();
  });

  test("should return empty column branches list", () => {
    const { groupBranches } = getGroupBranches(groups);
    const result = getColumnBranches(groupBranches as Record<string, GroupBranch>, columns, []);
    expect(result).toEqual([]);
  });

  test("should return column branches (only default branche)", () => {
    const { groupBranches } = getGroupBranches(groups);
    let result = getColumnBranches(groupBranches as Record<string, GroupBranch>, columns, [10, 11, 12]);
    expect(result).toMatchSnapshot();
    result = getColumnBranches(groupBranches as Record<string, GroupBranch>, {}, [1, 2, 3, 4]);
    expect(result).toMatchSnapshot();
  });

  test("should return visible merged column branches", () => {
    const { groupBranches } = getGroupBranches(groups);
    const result = getColumnBranches(groupBranches as Record<string, GroupBranch>, columns, [0, 1, 2, 3, 4, 5, 6]);
    const mergedBranches = getMergedBranches(result);
    expect(mergedBranches).toMatchSnapshot();
  });
});
