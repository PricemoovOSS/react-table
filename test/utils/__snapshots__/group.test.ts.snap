// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grou utils should return column branches (only default branche) 1`] = `
[
  {
    "id": "default-group-id-root-0",
    "label": "",
    "size": 1,
    "subGroups": [
      {
        "id": "default-group-id-root-1",
        "label": "",
        "size": 1,
        "subGroups": [
          {
            "id": "default-group-id",
            "label": "",
            "size": 1,
          },
        ],
      },
    ],
  },
  {
    "id": "default-group-id-root-0",
    "label": "",
    "size": 1,
    "subGroups": [
      {
        "id": "default-group-id-root-1",
        "label": "",
        "size": 1,
        "subGroups": [
          {
            "id": "default-group-id",
            "label": "",
            "size": 1,
          },
        ],
      },
    ],
  },
  {
    "id": "default-group-id-root-0",
    "label": "",
    "size": 1,
    "subGroups": [
      {
        "id": "default-group-id-root-1",
        "label": "",
        "size": 1,
        "subGroups": [
          {
            "id": "default-group-id",
            "label": "",
            "size": 1,
          },
        ],
      },
    ],
  },
]
`;

exports[`Grou utils should return column branches (only default branche) 2`] = `
[
  {
    "id": "default-group-id-root-0",
    "label": "",
    "size": 1,
    "subGroups": [
      {
        "id": "default-group-id-root-1",
        "label": "",
        "size": 1,
        "subGroups": [
          {
            "id": "default-group-id",
            "label": "",
            "size": 1,
          },
        ],
      },
    ],
  },
  {
    "id": "default-group-id-root-0",
    "label": "",
    "size": 1,
    "subGroups": [
      {
        "id": "default-group-id-root-1",
        "label": "",
        "size": 1,
        "subGroups": [
          {
            "id": "default-group-id",
            "label": "",
            "size": 1,
          },
        ],
      },
    ],
  },
  {
    "id": "default-group-id-root-0",
    "label": "",
    "size": 1,
    "subGroups": [
      {
        "id": "default-group-id-root-1",
        "label": "",
        "size": 1,
        "subGroups": [
          {
            "id": "default-group-id",
            "label": "",
            "size": 1,
          },
        ],
      },
    ],
  },
  {
    "id": "default-group-id-root-0",
    "label": "",
    "size": 1,
    "subGroups": [
      {
        "id": "default-group-id-root-1",
        "label": "",
        "size": 1,
        "subGroups": [
          {
            "id": "default-group-id",
            "label": "",
            "size": 1,
          },
        ],
      },
    ],
  },
]
`;

exports[`Grou utils should return not empty branches list 1`] = `
{
  "default-group-id": {
    "id": "default-group-id-root-0",
    "label": "",
    "size": 1,
    "subGroups": [
      {
        "id": "default-group-id-root-1",
        "label": "",
        "size": 1,
        "subGroups": [
          {
            "id": "default-group-id",
            "label": "",
            "size": 1,
          },
        ],
      },
    ],
  },
  "subgroup-1": {
    "id": "group1",
    "label": "Group 1",
    "size": 1,
    "subGroups": [
      {
        "id": "subgroup-1-root-1",
        "label": "Sub-Group 1.1",
        "size": 1,
        "subGroups": [
          {
            "id": "subgroup-1",
            "label": "",
            "size": 1,
          },
        ],
      },
    ],
  },
  "subgroup-2": {
    "id": "group2",
    "label": "Group 2",
    "size": 1,
    "subGroups": [
      {
        "id": "subgroup-2-root-1",
        "label": "Sub-Group 2.1",
        "size": 1,
        "subGroups": [
          {
            "id": "subgroup-2",
            "label": "",
            "size": 1,
          },
        ],
      },
    ],
  },
  "subgroup-4": {
    "id": "group2",
    "label": "Group 2",
    "size": 1,
    "subGroups": [
      {
        "id": "subgroup-3",
        "label": "Sub-Group 2.2",
        "size": 1,
        "subGroups": [
          {
            "id": "subgroup-4",
            "label": "Sub-Group 2.2.1",
            "size": 1,
          },
        ],
      },
    ],
  },
  "subgroup-5": {
    "id": "group2",
    "label": "Group 2",
    "size": 1,
    "subGroups": [
      {
        "id": "subgroup-3",
        "label": "Sub-Group 2.2",
        "size": 1,
        "subGroups": [
          {
            "id": "subgroup-5",
            "label": "Sub-Group 2.2.2",
            "size": 1,
          },
        ],
      },
    ],
  },
}
`;

exports[`Grou utils should return visible column branches 1`] = `
[
  {
    "id": "group1",
    "label": "Group 1",
    "size": 1,
    "subGroups": [
      {
        "id": "subgroup-1-root-1",
        "label": "Sub-Group 1.1",
        "size": 1,
        "subGroups": [
          {
            "id": "subgroup-1",
            "label": "",
            "size": 1,
          },
        ],
      },
    ],
  },
  {
    "id": "group1",
    "label": "Group 1",
    "size": 1,
    "subGroups": [
      {
        "id": "subgroup-1-root-1",
        "label": "Sub-Group 1.1",
        "size": 1,
        "subGroups": [
          {
            "id": "subgroup-1",
            "label": "",
            "size": 1,
          },
        ],
      },
    ],
  },
  {
    "id": "default-group-id-root-0",
    "label": "",
    "size": 1,
    "subGroups": [
      {
        "id": "default-group-id-root-1",
        "label": "",
        "size": 1,
        "subGroups": [
          {
            "id": "default-group-id",
            "label": "",
            "size": 1,
          },
        ],
      },
    ],
  },
  {
    "id": "group2",
    "label": "Group 2",
    "size": 1,
    "subGroups": [
      {
        "id": "subgroup-2-root-1",
        "label": "Sub-Group 2.1",
        "size": 1,
        "subGroups": [
          {
            "id": "subgroup-2",
            "label": "",
            "size": 1,
          },
        ],
      },
    ],
  },
  {
    "id": "group2",
    "label": "Group 2",
    "size": 1,
    "subGroups": [
      {
        "id": "subgroup-3",
        "label": "Sub-Group 2.2",
        "size": 1,
        "subGroups": [
          {
            "id": "subgroup-5",
            "label": "Sub-Group 2.2.2",
            "size": 1,
          },
        ],
      },
    ],
  },
  {
    "id": "group2",
    "label": "Group 2",
    "size": 1,
    "subGroups": [
      {
        "id": "subgroup-3",
        "label": "Sub-Group 2.2",
        "size": 1,
        "subGroups": [
          {
            "id": "subgroup-5",
            "label": "Sub-Group 2.2.2",
            "size": 1,
          },
        ],
      },
    ],
  },
  {
    "id": "default-group-id-root-0",
    "label": "",
    "size": 1,
    "subGroups": [
      {
        "id": "default-group-id-root-1",
        "label": "",
        "size": 1,
        "subGroups": [
          {
            "id": "default-group-id",
            "label": "",
            "size": 1,
          },
        ],
      },
    ],
  },
]
`;

exports[`Grou utils should return visible merged column branches 1`] = `
[
  {
    "id": "group1",
    "label": "Group 1",
    "size": 2,
    "subGroups": [
      {
        "id": "subgroup-1-root-1",
        "label": "Sub-Group 1.1",
        "size": 2,
        "subGroups": [
          {
            "id": "subgroup-1",
            "label": "",
            "size": 2,
          },
        ],
      },
    ],
  },
  {
    "id": "default-group-id-root-0",
    "label": "",
    "size": 1,
    "subGroups": [
      {
        "id": "default-group-id-root-1",
        "label": "",
        "size": 1,
        "subGroups": [
          {
            "id": "default-group-id",
            "label": "",
            "size": 1,
          },
        ],
      },
    ],
  },
  {
    "id": "group2",
    "label": "Group 2",
    "size": 3,
    "subGroups": [
      {
        "id": "subgroup-2-root-1",
        "label": "Sub-Group 2.1",
        "size": 1,
        "subGroups": [
          {
            "id": "subgroup-2",
            "label": "",
            "size": 1,
          },
        ],
      },
      {
        "id": "subgroup-3",
        "label": "Sub-Group 2.2",
        "size": 2,
        "subGroups": [
          {
            "id": "subgroup-5",
            "label": "Sub-Group 2.2.2",
            "size": 2,
          },
        ],
      },
    ],
  },
  {
    "id": "default-group-id-root-0",
    "label": "",
    "size": 1,
    "subGroups": [
      {
        "id": "default-group-id-root-1",
        "label": "",
        "size": 1,
        "subGroups": [
          {
            "id": "default-group-id",
            "label": "",
            "size": 1,
          },
        ],
      },
    ],
  },
]
`;
